pkg/resources/kafka/headlessService_test.go

```diff
@@ -0,0 +1,242 @@
+// limitations under the License.
+
+package kafka
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/banzaicloud/koperator/api/v1beta1"
+)
+
+func TestCheckIfNodePortSvcNeeded(t *testing.T) {
+	testReconciler := Reconciler{}
+
+	testCases := []struct {
+		testName            string
+		kafkaCluster        *v1beta1.KafkaCluster
+		isNodePortSvcNeeded bool
+	}{
+		{
+			testName: "KafkaCluster with all external listeners using NodePort",
+			kafkaCluster: &v1beta1.KafkaCluster{
+				Spec: v1beta1.KafkaClusterSpec{
+					ListenersConfig: v1beta1.ListenersConfig{
+						ExternalListeners: []v1beta1.ExternalListenerConfig{
+							{
+								AccessMethod: corev1.ServiceTypeNodePort,
+							},
+							{
+								AccessMethod: corev1.ServiceTypeNodePort,
+							},
+						},
+					},
+				},
+			},
+			isNodePortSvcNeeded: true,
+		},
+		{
+			testName: "KafkaCluster with all external listeners using non-NodePort",
+			kafkaCluster: &v1beta1.KafkaCluster{
+				Spec: v1beta1.KafkaClusterSpec{
+					ListenersConfig: v1beta1.ListenersConfig{
+						ExternalListeners: []v1beta1.ExternalListenerConfig{
+							{
+								AccessMethod: corev1.ServiceTypeLoadBalancer,
+							},
+							{
+								AccessMethod: corev1.ServiceTypeClusterIP,
+							},
+						},
+					},
+				},
+			},
+			isNodePortSvcNeeded: false,
+		},
+		{
+			testName: "KafkaCluster with external listeners of mixed usages of NodePort and non-NodePort",
+			kafkaCluster: &v1beta1.KafkaCluster{
+				Spec: v1beta1.KafkaClusterSpec{
+					ListenersConfig: v1beta1.ListenersConfig{
+						ExternalListeners: []v1beta1.ExternalListenerConfig{
+							{
+								AccessMethod: corev1.ServiceTypeLoadBalancer,
+							},
+							{
+								AccessMethod: corev1.ServiceTypeNodePort,
+							},
+						},
+					},
+				},
+			},
+			isNodePortSvcNeeded: true,
+		},
+	}
+
+	for _, test := range testCases {
+		testReconciler.KafkaCluster = test.kafkaCluster
+		got := testReconciler.checkIfNodePortSvcNeeded()
+		t.Run(test.testName, func(t *testing.T) {
+			if got != test.isNodePortSvcNeeded {
+				t.Errorf("Expected: %v  Got: %v", test.isNodePortSvcNeeded, got)
+			}
+		})
+	}
+}
+
+func TestGetNonNodePortSvc(t *testing.T) {
+	testCases := []struct {
+		testName         string
+		services         corev1.ServiceList
+		filteredServices corev1.ServiceList
+	}{
+		{
+			testName: "Services with mixed NodePort and non-NodePort services",
+			services: corev1.ServiceList{
+				Items: []corev1.Service{
+					{
+						ObjectMeta: metav1.ObjectMeta{
+							Name: "NodePort service 1",
+						},
+						Spec: corev1.ServiceSpec{
+							Type: corev1.ServiceTypeNodePort,
+						},
+					},
+					{
+						ObjectMeta: metav1.ObjectMeta{
+							Name: "NodePort service 2",
+						},
+						Spec: corev1.ServiceSpec{
+							Type: corev1.ServiceTypeNodePort,
+						},
+					},
```

Comment:
**Optional**: I **MAY** have considered moving one of node port services to a later point or maybe even create a 3rd one and have a `[firstNodePort, nonNodePort, middleNodePort, nonNodePort2, lastNodePort]` setup in the slice to make sure we handle items uniformly regardless of their position in the slice (and may have created a `[nonNodePort, middleNodePort, nonNodePort2]` case as well) because those logic errors in the function implementation always come unexpected when the initial implementation is modified and the logic is made more complex.
Not necessary here as we can pay attention to it the next time the function gets modified, just a highlight.